pyctcdecode/decoder.py

```diff
@@ -62,17 +63,70 @@
 # store frame information for each word, where frame is the logit index of (start_frame, end_frame)
 Frames = Tuple[int, int]
 WordFrames = Tuple[str, Frames]
-# all the beam information we need to keep track of during decoding
-# text, next_word, partial_word, last_char, text_frames, part_frames, logit_score
-Beam = Tuple[str, str, str, Optional[str], List[Frames], Frames, float]
-# same as BEAMS but with current lm score that will be discarded again after sorting
-LMBeam = Tuple[str, str, str, Optional[str], List[Frames], Frames, float, float]
-
-# for output beams we return the text, the scores, the lm state and the word frame indices
-# text, last_lm_state, text_frames, logit_score, lm_score
-OutputBeam = Tuple[str, Optional[AbstractLMState], List[WordFrames], float, float]
-# for multiprocessing we need to remove kenlm state since it can't be pickled
-OutputBeamMPSafe = Tuple[str, List[WordFrames], float, float]
+
+
+@dataclasses.dataclass(frozen=True)
+class Beam:
+    """Contains all the info needed for decoding a beam."""
+
+    text: str
+    next_word: str
+    partial_word: str
+    last_char: Optional[str]
+    text_frames: List[Frames]
+    partial_frames: Frames
+    logit_score: float
+
+    @classmethod
+    def from_lm_beam(cls, lm_beam: LMBeam) -> "Beam":
+        """Create a Beam from an LMBeam by stripping out the LM score."""
+        return Beam(
+            text=lm_beam.text,
+            next_word=lm_beam.next_word,
+            partial_word=lm_beam.partial_word,
+            last_char=lm_beam.last_char,
+            text_frames=lm_beam.text_frames,
+            partial_frames=lm_beam.partial_frames,
+            logit_score=lm_beam.logit_score,
+        )
+
+
+@dataclasses.dataclass(frozen=True)
+class LMBeam:
+    """Contains everything in a beam plus an additional language model score."""
+
+    text: str
+    next_word: str
+    partial_word: str
+    last_char: Optional[str]
+    text_frames: List[Frames]
+    partial_frames: Frames
+    logit_score: float
+    lm_score: float
+
+
+@dataclasses.dataclass(frozen=True)
+class OutputBeam:
+    """Output information."""
+
+    text: str
+    last_lm_state: Optional[AbstractLMState]
+    text_frames: List[WordFrames]
+    logit_score: float  # Cumulative logit score
+    lm_score: float  # Cumulative language model + logit score
+
+
+# May want to consider having an MP-safe LM state in case there are models that are safe to use
+@dataclasses.dataclass(frozen=True)
+class OutputBeamMPSafe:
+    """Output without LM state since that isn't necessarily multiprocessing-safe."""
+
+    text: str
+    text_frames: List[WordFrames]
+    logit_score: float  # Cumulative logit score
+    lm_score: float  # Cumulative language model + logit score
```

Comment:
Similarly, could we make `OutputBeam` a subclass of `OutputBeamMPSafe` that adds `last_lm_state`? Would likely need some renaming for it to make sense

Reply:
Here I would prefer to keep these separate for now - otherwise we'll lose the concept of a multiprocessing-safe vs unsafe. An alternative that I can think of is to have a single OutputBeam class with an optional state and then maybe the state object can have a `get_mp_safe_state()` function - this might generalize well for say a hypothetical language model that preserves its state in a way that works for multiprocessing, like if it saves a path (or relative position for some memory-mapped object) instead of a pointer

Reply:
I've just done what I mentioned as the other option and I think the end result works well here - it's cleaner and improves extensibility for the packages since we now don't have to be opinionated about whether or not model state is safe for multiprocessing